pkg/ddl/index_presplit.go

```diff
@@ -0,0 +1,417 @@
+		keys = append(keys, startKey)
+	}
+
+	// 2. Split in the end key.
+	endKey := tablecodec.EncodeTableIndexPrefix(physicalID, idxInfo.ID+1)
+	keys = append(keys, endKey)
+	return keys
+}
+
+func getSplitIdxKeysFromBound(
+	sctx sessionctx.Context,
+	tblInfo *model.TableInfo,
+	idxInfo *model.IndexInfo,
+	lower, upper []types.Datum,
+	splitNum int,
+) (keys [][]byte, err error) {
+	pi := tblInfo.GetPartitionInfo()
+	if pi == nil {
+		keys = make([][]byte, 0, splitNum)
+		return getSplitIdxPhysicalKeysFromBound(
+			sctx, tblInfo, idxInfo, tblInfo.ID, lower, upper, splitNum, keys)
+	}
+	keys = make([][]byte, 0, splitNum*len(pi.Definitions))
+	for _, p := range pi.Definitions {
+		keys, err = getSplitIdxPhysicalKeysFromBound(
+			sctx, tblInfo, idxInfo, p.ID, lower, upper, splitNum, keys)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return keys, nil
+}
+
+func getSplitIdxPhysicalKeysFromBound(
+	sctx sessionctx.Context,
+	tblInfo *model.TableInfo,
+	idxInfo *model.IndexInfo,
+	physicalID int64,
+	lower, upper []types.Datum,
+	splitNum int,
+	destKeys [][]byte,
+) ([][]byte, error) {
+	destKeys = getSplitIdxPhysicalStartAndOtherIdxKeys(tblInfo, idxInfo, physicalID, destKeys)
+	index := tables.NewIndex(physicalID, tblInfo, idxInfo)
+	// Split index regions by lower, upper value and calculate the step by (upper - lower)/num.
+	sc := sctx.GetSessionVars().StmtCtx
+	lowerIdxKey, _, err := index.GenIndexKey(sc.ErrCtx(), sc.TimeZone(), lower, kv.IntHandle(math.MinInt64), nil)
+	if err != nil {
+		return nil, err
+	}
+	// Use math.MinInt64 as handle_id for the upper index key to avoid affecting calculate split point.
+	// If use math.MaxInt64 here, test of `TestSplitIndex` will report error.
+	upperIdxKey, _, err := index.GenIndexKey(sc.ErrCtx(), sc.TimeZone(), upper, kv.IntHandle(math.MinInt64), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if bytes.Compare(lowerIdxKey, upperIdxKey) >= 0 {
+		lowerStr := datumSliceToString(lower)
+		upperStr := datumSliceToString(upper)
+		errMsg := fmt.Sprintf("Split index `%v` region lower value %v should less than the upper value %v",
+			idxInfo.Name, lowerStr, upperStr)
+		return nil, exeerrors.ErrInvalidSplitRegionRanges.GenWithStackByArgs(errMsg)
+	}
+	return util.GetValuesList(lowerIdxKey, upperIdxKey, splitNum, destKeys), nil
+}
+
+func datumSliceToString(ds []types.Datum) string {
+	str := "("
+	for i, d := range ds {
+		s, err := d.ToString()
+		if err != nil {
+			return fmt.Sprintf("%v", ds)
+		}
+		if i > 0 {
+			str += ","
+		}
+		str += s
+	}
+	str += ")"
+	return str
+}
+
+func splitIndexRegionAndWait(
+	ctx context.Context,
+	sctx sessionctx.Context,
+	store kv.Storage,
+	tblInfo *model.TableInfo,
+	idxInfo *model.IndexInfo,
+	splitIdxKeys [][]byte,
+) error {
+	s, ok := store.(kv.SplittableStore)
+	if !ok {
+		return nil
+	}
+	start := time.Now()
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, sctx.GetSessionVars().GetSplitRegionTimeout())
+	defer cancel()
```

Comment:
we need at least make sure split region is done, to avoid hot write, and we need retry on split failure. scatter region can wait at best, so maybe move this timeout for scatter only.

Reply:
If `SplitTable` is timed out, it will return an error and retry at caller side. It will not proceed to the next state.

Reply:
currently, this function doesn't return error

Reply:
d64f5840da596f